src/c++/vdo/base/vdo.c

```diff
@@ -477,6 +543,379 @@ static int register_vdo(struct vdo *vdo)
+		return result;
+
+	vdo->states.vdo.state = VDO_NEW;
+	return VDO_SUCCESS;
+}
+
+/**
+ * uninitialize_format_layout() - Unnitialize the format layout
+ *
+ * @vdo: The vdo to initialize.
+ *
+ * Return: VDO_SUCCESS or an error code.
+ **/
+static void uninitialize_format_layout(struct vdo *vdo)
+{
+	memset(&vdo->geometry, 0, sizeof(struct volume_geometry));
+	vdo_destroy_component_states(&vdo->states);
+}
+
+/**
+ * vdo_format() - Format a block device to function as a new VDO. This function
+ *                must be called on a device before a VDO can be loaded for the
+ *                first time. Once a device has been formatted, the VDO can be
+ *                loaded and shut down repeatedly. If a new VDO is desired, this
+ *                function should be called again.
+ *
+ * @vdo        The vdo to format.
+ * @error_ptr: The reason for any failure during this call.
+ *
+ * Return: VDO_SUCCESS or an error
+ **/
+static int __must_check vdo_format(struct vdo *vdo, char **error_ptr)
+{
+	int result;
+
+	result = initialize_format_layout(vdo, error_ptr);
+	if (result != VDO_SUCCESS)
+		return result;
+
+	result = write_format_layout(vdo, error_ptr);
+	uninitialize_format_layout(vdo);
+	return result;
+}
+
+/**
+ * vdo_is_formatted() - Check to see whether the areas vdo will write to during
+ *                      formatting are all zeroes.
+ *
+ * @vdo        The vdo to format.
+ * @formatted: The state of the device. 0 means ok to format, 1 otherwise.
+ *
+ * Return: VDO_SUCCESS or an error
+ **/
+static int __must_check vdo_is_formatted(struct vdo *vdo, int *formatted)
+{
+	char *block;
+	int result;
+	__le64 *data_le;
+	__le64 zero = __cpu_to_le64(0);
+
+	result = vdo_allocate(VDO_BLOCK_SIZE, u8, __func__, &block);
+	if (result != VDO_SUCCESS)
+		return result;
+
+	result = read_geometry_block(vdo, block);
+	if (result != VDO_SUCCESS) {
+		vdo_free(block);
+		return result;
+	}
+
+	// Following dm-thin-metadata.c logic
+	unsigned int block_size = VDO_BLOCK_SIZE / sizeof(__le64);
+
+	data_le = (__le64 *)block;
+	for (unsigned int i = 0; i < block_size; i++) {
+		if (data_le[i] != zero) {
+			*formatted = 1;
+			break;
+		}
+	}
+	vdo_free(block);
+	return VDO_SUCCESS;
+}
+
+/**
+ * compute_index_blocks() - Compute the size that the indexer will take up.
+ * @vdo: The vdo whose index size is calculated.
+ * @index_blocks_ptr: The number of blocks the index will use.
+ *
+ * Return: VDO_SUCCESS or an error code.
+ */
+static int compute_index_blocks(const struct device_config *config,
+				block_count_t *index_blocks_ptr)
+{
+	int result;
+	u64 index_bytes;
+	block_count_t index_blocks;
+
+	struct uds_parameters uds_parameters = {
+		.memory_size = config->index_memory,
+		.sparse = config->index_sparse,
+	};
+
+	result = uds_compute_index_size(&uds_parameters, &index_bytes);
+	if (result != UDS_SUCCESS) {
+		vdo_log_error("Error computing index size");
+		return -EINVAL;
+	}
+
+	index_blocks = index_bytes / VDO_BLOCK_SIZE;
```

Comment:
What is the point of this? If use isn't producing what we want, we should change UDS. If it is, there's no need to check it.